### PR TITLE
Type inference

### PR DESCRIPTION
**Pull request by Lipen**
This PR adds monumental work on the IFDS-based type inference engine.

### CODE REVIEW

**Comment by Lipen** [comment]
> Please, review detekt issues

@Saloed I have fixed some of them, but not all 300+.

File: usvm-dataflow-ts/src/main/kotlin/org/usvm/dataflow/ts/infer/EntryPointsProcessor.kt [hunk 1]
```diff
@@ -0,0 +1,25 @@
...
+        scene: EtsScene,
+    ): ArtificialMainWithAllMethods {
+        val artificialMainMethods = scene.projectClasses
+            .asSequence()
+            .flatMap { it.methods }
+            .filter { it.name == "@dummyMain" }
+            .toList()
+        return ArtificialMainWithAllMethods(
+            mainMethods = artificialMainMethods,
+            allMethods = scene.projectAndSdkClasses.flatMap { it.methods },
```
Comment (CaelmBleidd) [inline usvm-dataflow-ts/src/main/kotlin/org/usvm/dataflow/ts/infer/EntryPointsProcessor.kt:17]: Use project files only

File: usvm-dataflow-ts/src/main/kotlin/org/usvm/dataflow/ts/infer/TypeInferenceResult.kt [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+            },
+        )
+    }
+
+    private fun precalculateCaches(scene: EtsScene): Map<String, Set<EtsClass>> {
+        val result = hashMapOf<String, MutableSet<EtsClass>>()
+
+        scene.projectAndSdkClasses.forEach { clazz ->
+            clazz.methods.forEach {
+                result.computeIfAbsent(it.name) { hashSetOf() }.add(clazz)
```
Comment (CaelmBleidd) [inline usvm-dataflow-ts/src/main/kotlin/org/usvm/dataflow/ts/infer/TypeInferenceResult.kt:48]: Doesn't work for dflt

**Review by CaelmBleidd — CHANGES_REQUESTED** [review]

File: usvm-dataflow-ts/src/main/kotlin/org/usvm/dataflow/ts/infer/verify/EntityId.kt [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+) : EntityId {
+    constructor(local: EtsLocal, methodSignature: EtsMethodSignature)
+            : this(local.name, MethodId(methodSignature))
+}
+
+data class ThisId(
+    val method: MethodId
+) : EntityId {
+    constructor(methodSignature: EtsMethodSignature)
+            : this(MethodId(methodSignature))
```
[the remark this answers is not in the document]
Comment (Lipen) [reply usvm-dataflow-ts/src/main/kotlin/org/usvm/dataflow/ts/infer/verify/EntityId.kt]: Fixed all formatting

**Review by CaelmBleidd — APPROVED** [review]

**Pull request merged**